nethack/src/board.c

```diff
@@ -0,0 +1,226 @@
+               mdat->frameheight - 1, 2);
+    leaveok(gw->win, TRUE);
+    leaveok(mdat->content, TRUE);
+    wtimeout(gw->win, 500);
+}
+
+static void
+board_add_entry(struct nh_board_entry *entry, struct nh_menuitem **items,
+                int *size, int *icount, int maxwidth, int maxplayerwidth,
+                enum nh_menuitem_role txtrole)
+{
+    char fmtline[BUFSZ], line[BUFSZ], name[BUFSZ], hptxt[BUFSZ], entxt[BUFSZ];
+    struct tm tm;
+    time_t t, now;
+    char *inactive;
+
+    now = time(NULL);
+
+    strptime(entry->lastactive, "%Y-%m-%d %H:%M:%S.000000%z", &tm);
+    t = mktime(&tm);
+
+    if(difftime(now, t) > 30) { //Marks a player inactive after 30 seconds
+        inactive = "I";
+    } else {
+        inactive = "";
+    }
+
+    sprintf(name, "%s the %s %s %s %s", entry->name, entry->plalign,
+            entry->plgend, entry->plrace, entry->plrole);
+    sprintf(hptxt, "%d/%d", entry->hp, entry->hpmax);
+    sprintf(entxt, "%d/%d", entry->en, entry->enmax);
+    sprintf(fmtline, "%%1s  %%-%ds | %%-12s | %%-7s %%-7s | %%3d %%3d %%3d %%3d %%3d %%3d | %%5d | %%5d | %%5d",
+            maxplayerwidth);
+    sprintf(line, fmtline, inactive, name,
+            entry->leveldesc, hptxt, entxt, entry->wi, entry->in, entry->st,
+            entry->dx, entry->co, entry->ch, entry->moves, entry->depth,
+            entry->level);
+    add_menu_txt(*items, *size, *icount, line, txtrole);
+}
+
+static int
+get_max_playerdesc(struct nh_board_entry *entries, int listlen)
+{
+    int i, tmpsize, maxsize = 0;
+    struct nh_board_entry *tmpentry;
+    for(i = 0; i < listlen; i++) {
+        tmpentry = (entries + i);
+        tmpsize = strlen(tmpentry->name) //Length of the player's name
+                    + 5                  //Length of " the "
+                    + strlen(tmpentry->plalign) //Length of their alignment
+                    + 1                  //Space
+                    + strlen(tmpentry->plgend)
+                    + 1                  //Space
+                    + strlen(tmpentry->plrace)
+                    + 1                  //Space
+                    + strlen(tmpentry->plrole);
+        if(maxsize < tmpsize) {
+            maxsize = tmpsize;
+        }
+    }
+    return maxsize;
+}
+
+static void
+handle_key(int key)
+{
+    switch (key) {
+    case KEY_ESC:
+    case 'q':
+        done = TRUE;
+        draw_menu(gw);
+        break;
+    default:
+        break;
+    }
+}
+
+static void
+update_board(struct nh_menuitem **items, char *time)
+{
+    struct nh_board_entry *new_entries;
+    int i, new_listlen, new_maxplayersize;
+    int icount, size;
+    char fmtline[BUFSZ], line[BUFSZ];
+
+    new_entries = nh_get_board_entries(time, &new_listlen);
+
+    icount = 0;
+    size = 4 + listlen * 2;     /* maximum length, only required if every item
+                                   wraps */
+
+    free((*items));
+    (*items) = malloc(size * sizeof (struct nh_menuitem));
```

Comment:
*Do not do multiplication in a `malloc(3)` call*

Prefer `calloc(3)` or import `reallocarray(3)` from OpenBSD (it has an agreeable license).

Alternatively, check for integer overflow.